include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -180,89 +276,109 @@ namespace Utilities
                   }
               }
 
-            // protect the following communication steps using a mutex:
-            static CollectiveMutex      mutex;
-            CollectiveMutex::ScopedLock lock(mutex, comm);
-
-            const int mpi_tag =
-              Utilities::MPI::internal::Tags::dictionary_reinit;
-
             n_dict_procs_in_owned_indices = buffers.size();
             std::vector<MPI_Request> request;
             request.reserve(n_dict_procs_in_owned_indices);
 
-            // 3) send messages with local dofs to the right dict process
-            for (const auto &rank_pair : buffers)
-              {
-                request.push_back(MPI_Request());
-                const int ierr = MPI_Isend(rank_pair.second.data(),
-                                           rank_pair.second.size() * 2,
-                                           DEAL_II_DOF_INDEX_MPI_TYPE,
-                                           rank_pair.first,
-                                           mpi_tag,
-                                           comm,
-                                           &request.back());
-                AssertThrowMPI(ierr);
-              }
-
-            // 4) receive messages until all dofs in dict are processed
-            while (this->local_size != dic_local_received)
+            if (Utilities::MPI::sum(owned_indices.n_elements(), comm) ==
```

Comment:
Please add a comment here that explains that we switch between two implementations based on the check if every element is owned by exactly one rank.